packages/sitecore-jss-proxy/src/middleware/editing/render.ts

```diff
@@ -0,0 +1,143 @@
+import {
+  GraphQLEditingService,
+  EDITING_ALLOWED_ORIGINS,
+  RenderMetadataQueryParams,
+} from '@sitecore-jss/sitecore-jss/editing';
+
+/**
+ * Configuration for the editing render endpoint
+ */
+export type EditingRenderEndpointOptions = {
+  /**
+   * Custom path for the endpoint. Default is `<routerPath>/render`
+   * @example
+   * { path: '/foo/render' } -> <routerPath>/foo/render
+   */
+  path?: string;
+  /**
+   * GraphQl Request Client Factory provided by the server bundle
+   */
+  clientFactory: GraphQLRequestClientFactory;
+  /**
+   * The appRenderer will produce the requested route's html
+   */
+  renderView: AppRenderer;
+};
+
+type MetadataRequest = Request & { query: RenderMetadataQueryParams };
+
+/**
+ * Middleware to handle editing render requests
+ * @param {EditingRenderEndpointOptions} config for the endpoint
+ */
+export const editingRenderMiddleware = (config: EditingRenderEndpointOptions) => async (
+  req: MetadataRequest,
+  res: Response
+): Promise<void> => {
+  try {
+    debug.editing('editing render middleware start');
+
+    const startTimestamp = Date.now();
+
+    const query = req.query;
+
+    const requiredQueryParams: (keyof RenderMetadataQueryParams)[] = [
+      'sc_site',
+      'sc_itemid',
+      'sc_lang',
+      'route',
+      'mode',
+    ];
+
+    const missingQueryParams = requiredQueryParams.filter((param) => !query[param]);
+
+    // Validate query parameters
+    if (missingQueryParams.length) {
+      debug.editing('missing required query parameters: %o', missingQueryParams);
+
+      res.status(400).send(`Missing required query parameters: ${missingQueryParams.join(', ')}`);
+
+      return;
+    }
+
+    const graphQLEditingService = new GraphQLEditingService({
+      clientFactory: config.clientFactory,
+    });
+
+    const data = await graphQLEditingService.fetchEditingData({
+      siteName: query.sc_site,
+      itemId: query.sc_itemid,
+      language: query.sc_lang,
+      version: query.sc_version,
+      layoutKind: query.sc_layoutKind,
+    });
+
+    if (!data || !data.layoutData || !data.dictionary) {
+      throw new Error(`Unable to fetch editing data for ${JSON.stringify(query)}`);
+    }
+
+    const viewBag = { dictionary: data.dictionary };
+
+    config.renderView(
+      (err: Error | null, result: RenderResponse | null) => {
+        if (err) {
+          handleError(res, err);
+          return;
+        }
+
+        if (!result) {
+          debug.editing('editing render middleware end in %dms: %o', Date.now() - startTimestamp, {
+            status: 204,
+            route: query.route,
+          });
+
+          res.status(204).send();
+          return;
+        }
+
+        const statusCode = data.layoutData.sitecore.route ? 200 : 404;
+
+        // Restrict the page to be rendered only within the allowed origins
+        res.setHeader('Content-Security-Policy', getSCPHeader());
```

Comment:
all the cookies/headers from Pages should be propagated in addition to this one. Is it something the middleware will do implicitly?
Either way, perhaps a test should be added for this?

Reply:
Same thing for propagating query string parameters coming from Pages.

Reply:
hm not sure what you mean.. in nextjs middleware we have a specific issue and that's why we are modifying the cookie. do we need to do the same here?

Reply:
We don't only propagate cookies, but headers and query string params in general
```
      // Get query string parameters to propagate on subsequent requests (e.g. for deployment protection bypass)
      const params = this.getQueryParamsForPropagation(query);

      // Get headers to propagate on subsequent requests
      const headers = this.getHeadersForPropagation(incomingHeaders);
```

Originally this was done to support vercel protection (I think), but it's also useful for any Sitecore protection logic that is present now or may be introduced in future.

Reply:
@art-alexeyenko I discussed that and verified that we don't need to pass any headers/cookies, the same like for Metadata integration. We propagated headers for "Chromes" integration.
Now we don't have an internal request to RH, so we are good

Reply:
@art-alexeyenko All this stuff was done in order to execute request to RH, it's not applicable anymore